File: frontend/src/types/thesis.ts
export interface Student {
  id: number;
  name: string;
  graduation_year: number;
  thesis_title: string;
  thesis_summary: string;
  python_code?: string;
  created_at: string;
  updated_at?: string;
}

export interface SearchResponse {
  students: Student[];
  total: number;
  page: number;
  per_page: number;
  total_pages: number;
}

export interface ExecutionRequest {
  args?: string;
}

export interface ExecutionResult {
  success: boolean;
  result?: string;
  error?: string;
  execution_time?: number;
}

export interface ExecutionLog {
  id: number;
  student_id: number;
  input_args?: string;
  output_result?: string;
  success: boolean;
  error_message?: string;
  executed_at: string;
}

File: backend/app/schemas/thesis.py
from pydantic import BaseModel, Field
from typing import Optional
from datetime import datetime


class StudentBase(BaseModel):
    name: str = Field(..., min_length=1, max_length=255)
    graduation_year: int = Field(..., ge=1900, le=2100)
    thesis_title: str = Field(..., min_length=1, max_length=500)
    thesis_summary: str = Field(..., min_length=1)
    python_code: Optional[str] = None


class StudentCreate(StudentBase):
    pass


class StudentUpdate(BaseModel):
    name: Optional[str] = Field(None, min_length=1, max_length=255)
    graduation_year: Optional[int] = Field(None, ge=1900, le=2100)
    thesis_title: Optional[str] = Field(None, min_length=1, max_length=500)
    thesis_summary: Optional[str] = Field(None, min_length=1)
    python_code: Optional[str] = None


class Student(StudentBase):
    id: int
    created_at: datetime
    updated_at: Optional[datetime] = None

    class Config:
        from_attributes = True


class ExecutionRequest(BaseModel):
    args: Optional[str] = None


class ExecutionResult(BaseModel):
    success: bool
    result: Optional[str] = None
    error: Optional[str] = None
    execution_time: Optional[float] = None


class ExecutionLogBase(BaseModel):
    student_id: int
    input_args: Optional[str] = None
    output_result: Optional[str] = None
    success: bool = False
    error_message: Optional[str] = None


class ExecutionLog(ExecutionLogBase):
    id: int
    executed_at: datetime

    class Config:
        from_attributes = True


class SearchResponse(BaseModel):
    students: list[Student]
    total: int
    page: int
    per_page: int
    total_pages: int